protocol/morse/sanctioned_endpoints_store.go

```diff
@@ -0,0 +1,190 @@
+
+			// Apply the appropriate type of sanction
+			switch recommendedSanction {
+
+			// Permanent sanction: these persist for the lifetime of the process but are not persisted to disk.
+			// They will be lost on PATH process restart and are not shared across multiple instances.
+			case protocolobservations.MorseSanctionType_MORSE_SANCTION_PERMANENT:
+				logger.Info().Msg("Adding permanent sanction for endpoint")
+
+				ses.addPermanentSanction(
+					protocol.EndpointAddr(endpointObservation.GetEndpointAddr()),
+					sanctionData,
+				)
+
+			// Session-based sanctions: These automatically expire after a set duration.
+			case protocolobservations.MorseSanctionType_MORSE_SANCTION_SESSION:
+				logger.Info().Msg("Adding session sanction for endpoint")
+
+				ses.addSessionSanction(
+					protocol.EndpointAddr(endpointObservation.GetEndpointAddr()),
+					endpointObservation.GetAppAddress(),
+					endpointObservation.GetSessionKey(),
+					sanctionData,
+				)
+			default:
+				logger.Warn().Msg("sanction type not supported by the store. skipping.")
+			}
+		}
+	}
+}
+
+// FilterSanctionedEndpoints removes sanctioned endpoints from the provided list.
+// It returns a filtered list containing only endpoints that don't have active sanctions.
+// This method is used during endpoint selection to avoid selecting endpoints that have been sanctioned.
+func (ses *sanctionedEndpointsStore) FilterSanctionedEndpoints(
+	allEndpoints []endpoint,
+	appAddr string,
+	sessionKey string,
+) []endpoint {
+	var filteredEndpoints []endpoint
+
+	for _, endpoint := range allEndpoints {
+		sanctioned, reason := ses.isSanctioned(endpoint.Addr(), appAddr, sessionKey)
+		if sanctioned {
+			logger := loggerWithEndpoint(ses.logger, appAddr, sessionKey, endpoint.Addr(), reason)
+			logger.Debug().Msg("Filtering out sanctioned endpoint")
+			continue
+		}
+		filteredEndpoints = append(filteredEndpoints, endpoint)
+	}
+
+	return filteredEndpoints
+}
+
+// addPermanentSanction adds a permanent sanction for an endpoint.
+// Permanent sanctions never expire and require manual intervention to remove.
+// These are used for serious errors like validation failures that may indicate malicious behavior.
+func (ses *sanctionedEndpointsStore) addPermanentSanction(
+	endpointAddr protocol.EndpointAddr,
+	sanctionData sanction,
+) {
+	ses.mu.Lock()
+	ses.permanentSanctions[endpointAddr] = sanctionData
+	ses.mu.Unlock()
+}
+
+// addSessionSanction adds a session-based sanction for an endpoint.
+// Session-based sanctions automatically expire after defaultSessionSanctionExpiration.
+// These are used for temporary issues like timeouts or connection problems.
+func (ses *sanctionedEndpointsStore) addSessionSanction(
+	endpointAddr protocol.EndpointAddr,
+	appAddr string,
+	sessionKey string,
+	sanctionData sanction,
+) {
+	// TODO_MVP(@adshmh): validate the session key format. Should not create a sanction entry if the session key is not valid.
```

Comment:
Can you move this TODO to `sessionSanctionKey` and update it to saying "return an error if any of the composite key components are empty"?